tools/cmake/mbed_python_interpreter.cmake

```
@@ -4,32 +4,63 @@
 
+get_filename_component(MBED_CE_TOOLS_BASE_DIR "${CMAKE_CURRENT_LIST_DIR}/.." ABSOLUTE)
+
 if(MBED_CREATE_PYTHON_VENV)
     # Use the venv.
 
     # Note: venv is stored in the source directory as it can be shared between all the build directories
     # (not target specific)
     set(MBED_VENV_LOCATION ${MBED_SOURCE_DIR}/venv)
     set(VENV_STAMP_FILE ${MBED_VENV_LOCATION}/mbed-venv.stamp)
-    set(MBED_REQUIREMENTS_TXT_LOCATION "${CMAKE_CURRENT_LIST_DIR}/../requirements.txt")
+    set(MBED_PYPROJECT_TOML_LOCATION "${MBED_CE_TOOLS_BASE_DIR}/pyproject.toml")
 
-    # Make it so modifying requirements.txt will trigger a reconfigure
-    set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${MBED_REQUIREMENTS_TXT_LOCATION})
+    # Make it so modifying pyproject.toml will trigger a reconfigure
+    set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${MBED_PYPROJECT_TOML_LOCATION})
 
-    # Find Python3, using the venv if it already exists
-    set (ENV{VIRTUAL_ENV} ${MBED_VENV_LOCATION})
-    set (Python3_FIND_VIRTUALENV FIRST)
-    find_package(Python3 REQUIRED COMPONENTS Interpreter)
+    # Find Python3 (this will get the one in the venv if we already found it)
+    set(ENV{VIRTUAL_ENV} ${MBED_VENV_LOCATION})
+    set(Python3_FIND_VIRTUALENV FIRST)
+    find_package(Python3 COMPONENTS Interpreter)
     include(CheckPythonPackage)
 
     set(NEED_TO_CREATE_VENV FALSE)
     set(NEED_TO_INSTALL_PACKAGES FALSE)
-    if(NOT EXISTS "${VENV_STAMP_FILE}")
+
+    # Special situation: if we have a cached interpreter location in the venv dir, but Python could be found,
+    # this means that the venv was deleted or symlinks to a missing python install location.
+    # So, use the system python and recreate it.
+    if("${Python3_EXECUTABLE}" MATCHES "${MBED_VENV_LOCATION}" AND NOT Python3_FOUND)
```

I have noticed two scenarios here that seem to cause problems:

1. The python venv is corrupted in some way (e.g. by removal of the system Python that it is pointing to). This causes `find_package(Python3)` to fail, with an error like "Python interpreter could not be run"
2. The python venv is deleted, which will cause `find_package(Python3)` to silently return the system python instead of the venv.

I attempt to handle both those cases here. I have confirmed that #2 is fixed now, but I have not found a reliable way to reproduce situation #1 yet, so I sort of threw everything at it in the hope that we can make FindPython temporarily revert back to the system python.